### PR TITLE
check_source.py: Validate content of _links in submissions

### PR DESCRIPTION
**Pull request by Vogtinator**
Recursively expand links until the end or reaching a "safe" project.
When a link isn't static, fail the review.

Draft because:
- [ ] Breaks submissions to devel projects, because branches to them aren't allowed
- [ ] Allowing links to different packages in `openSUSE:` might also be bad
- [ ] Needs to check for project links as well?
- [ ] The `_link` parsing needs special care e.g. `package="../evil/package"` (`osc.core.makeurl` just does `'/'.join(path)`!)

### CODE REVIEW

File: check_source.py [hunk 1]
```diff
@@ -114,8 +148,8 @@ def check_source_submission(self, source_project, source_package, source_revisio
             self.review_messages['declined'] = 'Only one action per request allowed'
             return False
 
-        if source_revision is None:
-            self.review_messages['declined'] = 'Submission not from a pinned source revision'
+        # The SR itself can use the commit number safely, because it's declined on package deletion
```
Comment (DimStar77) [inline check_source.py:151]: but after recreating the package, the SR can be 'reopened'

Comment (Vogtinator) [reply check_source.py:151]: IIRC I tried this once but got `is in a final state and can't be changed`

Comment (Vogtinator) [reply check_source.py:151]: If we have to enforce hashes in SRs, then pretty much every SR to `oS:F` currently will be declined...

Comment (Vogtinator) [reply check_source.py:151]: Reopening should also restart the reviews. But if it happens quickly enough, then the caches might make this a false positive.

(This rabbit hole goes really deep...)

Comment (DimStar77) [reply check_source.py:151]: https://build.opensuse.org/request/show/897561

Comment (DimStar77) [reply check_source.py:151]: declining non-hash SRs can't be enforced without the backend transparently changing the rev - a simple `osc sr DEVELPRJ PKG openSUSE:Factory` creates an sr with a numeric rev

Comment (Vogtinator) [reply check_source.py]: For a transition period, the bot could decline the sr and create a new one with hash instead

Comment (DimStar77) [reply check_source.py]: giving the users a decline message on everything and then not having the users as creator makes it hard to enter into dialogs through the SR UI

Should rather be the backend rewriting it from 'rev x' to hash-rev (it has all the information after all)

**Comment by coolo** [comment]
Why so complicated? You only want to verify it's expanded sources on a pinned revision, no? If you have a revision, compare the srcmd5 of /source/$project/$package?rev=$rev with /source/$project/$package?rev=$rev&expand=1

I don't see why it's important where the sources point to - as long as we accept what we reviewed.

File: check_source.py (hunk 1 above)
Comment (coolo) [reply check_source.py]: The above SR just proves that we shouldn't try to enforce this client side. This really should be enforced on the source revision system

Comment (Vogtinator) [reply check_source.py]: Yes, so far there's no progress or even acknowledgement on the OBS side.

Comment (coolo) [reply check_source.py]: Yeah, let me use my levers... :)

**Comment by coolo** [comment]
Moving to https://github.com/openSUSE/open-build-service/pull/11263

**Pull request closed without merging**